File: practica_1/Practica1.0.c
/*
 * Programa: Generador de Personajes RPG
 * Autor: Carlos Mendoza
 * Versión: 2.0.0
 * Descripción: Sistema completo con menú interactivo.
 * Fecha: 2026-03-25
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

void limpiar_buffer() {
    int c;
    while ((c = getchar()) != '\n' && c != EOF);
}

const char *atributos[] = {
    "Fuerza", "Destreza", "Constitución",
    "Inteligencia", "Sabiduría", "Carisma"
};

int calcular_modificador(int valor) {
    return (valor - 10) / 2;
}

void generar_atributos(int stats[6]) {
    for (int i = 0; i < 6; i++) {
        stats[i] = (rand() % 16) + 3;
    }
}

void mostrar_personaje(const char *nombre, int stats[6]) {
    printf("\n--- Hoja de Personaje: %s ---\n", nombre);
    for (int i = 0; i < 6; i++) {
        int mod = calcular_modificador(stats[i]);
        printf("%-14s: %2d (%+d)\n", atributos[i], stats[i], mod);
    }
    printf("----------------------------------\n\n");
}

void crear_personaje() {
    char nombre[50];
    printf("\nIntroduce el nombre de tu héroe: ");
    getchar();
    fgets(nombre, sizeof(nombre), stdin);

    for (int i = 0; nombre[i]; i++) {
        if (nombre[i] == '\n') { 
            nombre[i] = '\0'; 
            break; 
        }
    }

    if (nombre[0] != '\0') {
        int stats[6];
        generar_atributos(stats);
        mostrar_personaje(nombre, stats);
    } else {
        printf("Error: El personaje necesita un nombre para existir.\n");
    }
}

void mostrar_menu() {
    printf("\n=== GENERADOR DE PERSONAJES RPG ===\n");
    printf("1. Crear nuevo personaje\n");
    printf("2. Acerca de\n");
    printf("3. Salir\n");
    printf("Selecciona una opción: ");
}

int main() {
    srand(time(NULL));
    int opcion;

    printf("¡Bienvenido al Generador de Personajes RPG!\n");

    do {
        mostrar_menu();
    if (scanf("%d", &opcion) != 1) {
    printf("\nEntrada inválida.\n");
    limpiar_buffer();
    opcion = 0;
    continue;
}

        switch(opcion) {
            case 1:
                crear_personaje();
                break;
            case 2:
                printf("\nGenerador de Personajes RPG v2.0.0\n");
                printf("Creado para práctica de Sistemas Operativos\n\n");
                break;
            case 3:
                printf("\n¡Hasta pronto, aventurero!\n");
                break;
            default:
                printf("\nOpción no válida. Intenta de nuevo.\n");
        }
    } while(opcion != 3);

    return 0;
}

   

File: practica_1/practica1.0.py
"""
Programa: Generador de Personajes RPG
Autor: Equipo de programación
Versión: 1.3.0
Descripción: Generador con sistema de razas y bonificaciones.
Fecha: 2026-03-27
"""

import random

RAZAS = {
    "Humano": {"Fuerza": 1, "Destreza": 1, "Constitución": 1, 
               "Inteligencia": 1, "Sabiduría": 1, "Carisma": 1},
    "Elfo": {"Destreza": 2, "Inteligencia": 1},
    "Enano": {"Constitución": 2, "Fuerza": 1},
    "Orco": {"Fuerza": 2, "Constitución": 1}
}

def calcular_modificador(valor):
    return (valor - 10) // 2

def generar_atributos():
    atributos = ["Fuerza", "Destreza", "Constitución", "Inteligencia", "Sabiduría", "Carisma"]
    stats = {atrib: random.randint(3, 18) for atrib in atributos}
    return stats

def aplicar_raza(stats, raza):
    bonificaciones = RAZAS.get(raza, {})
    for atrib, bonus in bonificaciones.items():
        stats[atrib] = stats.get(atrib, 0) + bonus
    return stats

def mostrar_personaje(nombre, stats, raza):
    print(f"\n--- Hoja de Personaje: {nombre} ({raza}) ---")
    for k, v in stats.items():
        mod = calcular_modificador(v)
        print(f"{k:<14}: {v:2d} ({mod:+d})")
    print("----------------------------------\n")
def main():
    print("¡Bienvenido al Generador de Personajes RPG!")
    
    while True:
        nombre = input("\nIntroduce el nombre de tu héroe: ")
        
        if not nombre:
            print("Error: El personaje necesita un nombre para existir.")
            continue
        
        print("\nRazas disponibles:")
        for i, raza in enumerate(RAZAS.keys(), 1):
            print(f"{i}. {raza}")
        
        try:
            opcion = int(input("Selecciona una raza: "))
            raza = list(RAZAS.keys())[opcion - 1]
        except (ValueError, IndexError):
            print("Opción inválida. Se usará Humano por defecto.")
            raza = "Humano"
        
        mis_stats = generar_atributos()
        mis_stats = aplicar_raza(mis_stats, raza)
        mostrar_personaje(nombre, mis_stats, raza)

        repetir = input("¿Quieres crear otro personaje? (s/n): ").lower()
        if repetir != 's':
            print("¡Hasta pronto, aventurero!")
            break
